refactor(semantic): change the reference flag to `ReferenceFlags::Type` if it is used within a `TSTypeQuery`

Dunqing opened the pull request:
So far, the `ReferenceFlags::TSTypeQuery` only used indicates it is referenced by `TSTypeQuery` that we can confirm the reference should be regarded as a type reference, namely `ReferenceFlags::Type`.

This PR adds a `ReferenceFlags::ValueAsType` instead of `ReferenceFlags::TSTypeQuery`.  The new flag has the same behavior as the previous one. But it looks more general and is not only used in `TSTypeQuery`. But now it is a temporary flag. We use it to resolve the symbol correctly and replace `ReferenceFlags::ValueAsTyoe` with `ReferenceFlags::Type` after resolved. 

Also, this change eliminates the inconsistency in behavior between the `Reference::is_type` and `ReferenceFlags::is_type` methods.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5444?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#5444** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5444?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#5441** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5441?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac commented on crates/oxc_semantic/src/reference.rs:
```diff
@@ -117,6 +117,6 @@ impl Reference {
     #[inline]
     pub fn is_type(&self) -> bool {
-        self.flags.is_type() || self.flags.is_ts_type_query()
+        self.flags.is_type()
```
this behavior is depended upon by `no-unused-vars`. If we want to change what `is_type` does, we need to re-add this API under a different name.

DonIsaac replied:
maybe we should have an `is_used_as_type` method?

Dunqing replied:
The behavior has not changed

 `ReferenceFlags` that used to be `ReferenceFlags::Read` | `ReferenceFlags::TSTypeQuery` are now `ReferenceFlags::Type`.

DonIsaac replied:
Nothing fails = happy Don

Dunqing committed:
refactor(semantic): change the reference flag to `ReferenceFlags::Type` if it is used within a `TSTypeQuery` (#5444)

So far, the `ReferenceFlags::TSTypeQuery` only used indicates it is referenced by `TSTypeQuery` that we can confirm the reference should be regarded as a type reference, namely `ReferenceFlags::Type`.

This PR adds a `ReferenceFlags::ValueAsType` instead of `ReferenceFlags::TSTypeQuery`.  The new flag has the same behavior as the previous one. But it looks more general and is not only used in `TSTypeQuery`. But now it is a temporary flag. We use it to resolve the symbol correctly and replace `ReferenceFlags::ValueAsTyoe` with `ReferenceFlags::Type` after resolved.

Also, this change eliminates the inconsistency in behavior between the `Reference::is_type` and `ReferenceFlags::is_type` methods.
